### PR TITLE
ARROW-5742: [CI][C++] Add nightly Valgrind build

### PR DESCRIPTION
**Pull request by pitrou**

### CODE REVIEW

**Comment by pitrou** [comment]
@kszucs

**Comment by pitrou** [comment]
I've checked this runs successfully locally (though slowly):
```
$ docker-compose run --rm conda-cpp-valgrind

[ ... ]

100% tests passed, 0 tests failed out of 78

Label Time Summary:
arrow-tests      = 829.10 sec*proc (60 tests)
arrow_dataset    =  96.69 sec*proc (8 tests)
parquet-tests    = 350.21 sec*proc (7 tests)
plasma-tests     = 147.54 sec*proc (3 tests)
unittest         = 1423.54 sec*proc (78 tests)

Total Test time (real) = 187.49 sec
+ '[' '' == ON ']'
+ popd
/
```

**Comment by pitrou** [comment]
@ursabot crossbow build test-conda-cpp-valgrind

**Comment by pitrou** [comment]
@ursabot crossbow help

**Comment by pitrou** [comment]
@ursabot crossbow test-conda-cpp-valgrind

**Comment by pitrou** [comment]
@ursabot crossbow test-conda-cpp-valgrind

**Comment by pitrou** [comment]
@ursabot crossbow run test-conda-cpp-valgrind

**Comment by pitrou** [comment]
@ursabot crossbow submit test-conda-cpp-valgrind

**Comment by kszucs** [comment]
[AMD64 Conda Crossbow Submit (#88852)](https://ci.ursalabs.org/#builders/98/builds/491) builder has been succeeded.

Revision: 43068985bf6a187ccc6ecf7ff3b6a035d9e93ad2

Submitted crossbow builds: [ursa-labs/crossbow @ ursabot-474](https://github.com/ursa-labs/crossbow/branches/all?query=ursabot-474)

|Task|Status|
|----|------|
|test-conda-cpp-valgrind|[CircleCI](https://circleci.com/gh/ursa-labs/crossbow/tree/ursabot-474-circle-test-conda-cpp-valgrind)|

**Comment by pitrou** [comment]
I've enabled some optimizations, which makes the build faster locally:
```
$ docker-compose run --rm conda-cpp-valgrind

[ ... ]

100% tests passed, 0 tests failed out of 78

Label Time Summary:
arrow-tests      = 325.75 sec*proc (60 tests)
arrow_dataset    =  44.34 sec*proc (8 tests)
parquet-tests    = 106.69 sec*proc (7 tests)
plasma-tests     =  43.06 sec*proc (3 tests)
unittest         = 519.84 sec*proc (78 tests)

```

**Comment by pitrou** [comment]
@ursabot crossbow submit test-conda-cpp-valgrind

**Comment by kszucs** [comment]
[AMD64 Conda Crossbow Submit (#88862)](https://ci.ursalabs.org/#builders/98/builds/492) builder has been succeeded.

Revision: 79d47a33bd1eb3d96cdbc23c4f7b2236bfef7ded

Submitted crossbow builds: [ursa-labs/crossbow @ ursabot-475](https://github.com/ursa-labs/crossbow/branches/all?query=ursabot-475)

|Task|Status|
|----|------|
|test-conda-cpp-valgrind|[CircleCI](https://circleci.com/gh/ursa-labs/crossbow/tree/ursabot-475-circle-test-conda-cpp-valgrind)|

File: docker-compose.yml [hunk 1]
```diff
@@ -85,13 +85,44 @@ services:
...
       - .:/arrow:delegated
       - ${ARCH}-conda-cache:/build:delegated
     command: &cpp-conda-command
       ["/arrow/ci/scripts/cpp_build.sh /arrow /build &&
         /arrow/ci/scripts/cpp_test.sh /arrow /build"]
 
+  conda-cpp-valgrind:
+    # Usage:
+    #   docker-compose build conda-cpp
+    #   docker-compose run --rm conda-cpp
```
Comment (kszucs) [inline docker-compose.yml]: `docker-compose run --rm conda-cpp-valgrind`

File: ci/scripts/cpp_build.sh [hunk 2]
```diff
@@ -98,7 +100,7 @@ cmake -G "${CMAKE_GENERATOR:-Ninja}" \
       -Dbenchmark_SOURCE=${benchmark_SOURCE:-AUTO} \
       -DBOOST_SOURCE=${BOOST_SOURCE:-AUTO} \
       -DBrotli_SOURCE=${Brotli_SOURCE:-AUTO} \
-      -DBUILD_WARNING_LEVEL=${DARROW_BUILD_WARNING_LEVEL:-CHECKIN} \
+      -DBUILD_WARNING_LEVEL=${BUILD_WARNING_LEVEL:-CHECKIN} \
```
Comment (kszucs) [inline ci/scripts/cpp_build.sh]: Hard to spot!

**Comment by kszucs** [comment]
@ursabot crossbow --help

**Comment by kszucs** [comment]
@ursabot crossbow submit --help

**Review by kszucs — CHANGES_REQUESTED** [review]
Just a nit in the docker-compose comment.

**Comment by pitrou** [comment]
@kszucs Anything else? macOS failure is unrelated.

**Review by kszucs — APPROVED** [review]
LGTM

**Pull request closed without merging**